### PR TITLE
Add reverse flag in associative scan

### PR DESCRIPTION
**Pull request by AdrienCorenflos**
Useful for some reverse mode associative accumulation.

Adrien

### CODE REVIEW

**Comment by AdrienCorenflos** [comment]
The failure looks to be related to the testing framework, not the code itself.

File: jax/lax/lax_control_flow.py [hunk 1]
```diff
@@ -2357,9 +2359,18 @@ def associative_scan(fn, elems):
...
+  Example 3: reversed partial sums of an array of numbers
+
+  >>> lax.associative_scan(jnp.add, jnp.arange(0, 4), reverse=True)
+  [ 6, 6, 5, 3]
   """
+  import jax.numpy as jnp
   elems_flat, tree = tree_flatten(elems)
 
+  if reverse:
+    elems_flat = [jnp.flip(elem, 0) for elem in elems_flat]
```
Comment (mattjj) [inline jax/lax/lax_control_flow.py:2372]:
```suggestion
    elems_flat = [lax.rev(elem, [0]) for elem in elems_flat]
```

File: jax/lax/lax_control_flow.py [hunk 2]
```diff
@@ -2357,9 +2359,18 @@ def associative_scan(fn, elems):
   >>> partial_prods = lax.associative_scan(jnp.matmul, mats)
   >>> partial_prods.shape
   (4, 2, 2)
+
+  Example 3: reversed partial sums of an array of numbers
+
+  >>> lax.associative_scan(jnp.add, jnp.arange(0, 4), reverse=True)
+  [ 6, 6, 5, 3]
   """
+  import jax.numpy as jnp
```
Comment (mattjj) [inline jax/lax/lax_control_flow.py:2368]:
```suggestion
```

File: jax/lax/lax_control_flow.py [hunk 3]
```diff
@@ -2434,6 +2445,9 @@ def _scan(elems):
 
   scans = _scan(elems_flat)
 
+  if reverse:
+    scans = [jnp.flip(scanned, 0) for scanned in scans]
```
Comment (mattjj) [inline jax/lax/lax_control_flow.py:2449]:
```suggestion
    scans = [lax.rev(scanned, [0]) for scanned in scans]
```

**Review by mattjj — COMMENTED** [review]
Thanks for contributing this! I suggest to use `lax.rev` to avoid the `jnp` import.

**Comment by mattjj** [comment]
I think the test failures may be existing failures on the master branch. You may want to merge/rebase and try again.

**Comment by mattjj** [comment]
Yes! Thanks for the ping.

**Review by mattjj — APPROVED** [review]
LGTM!

**Pull request merged**